tests/integration/index/query_with_composite_index_only_filter_test.go

```diff
@@ -0,0 +1,767 @@
+}
+
+func TestQueryWithCompositeIndex_WithLessOrEqualFilterOnSecondField_ShouldFetch(t *testing.T) {
+	req := `query {
+		User(filter: {age: {_le: 28}, name: {_ne: "Bruno"}}) {
+			name
+		}
+	}`
+	test := testUtils.TestCase{
+		Description: "Test index filtering with _le filter",
+		Actions: []any{
+			testUtils.SchemaUpdate{
+				Schema: schemaWithNameAgeIndex,
+			},
+			testUtils.CreatePredefinedDocs{
+				Docs: getUserDocs(),
+			},
+			testUtils.Request{
+				Request: req,
+				Results: []map[string]any{
+					{"name": "Fred"},
+					{"name": "Shahzad"},
+				},
+			},
+			testUtils.Request{
+				Request:  makeExplainQuery(req),
+				Asserter: testUtils.NewExplainAsserter().WithFieldFetches(2).WithIndexFetches(10),
+			},
+		},
+	}
+
+	testUtils.ExecuteTestCase(t, test)
+}
+
+func TestQueryWithCompositeIndex_WithNotEqualFilter_ShouldFetch(t *testing.T) {
+	req := `query {
+		User(filter: {name: {_ne: "Islam"}, age: {_ne: 28}}) {
+			name
+		}
+	}`
+	test := testUtils.TestCase{
+		Description: "Test index filtering with _ne filter",
+		Actions: []any{
+			testUtils.SchemaUpdate{
+				Schema: schemaWithNameAgeIndex,
+			},
+			testUtils.CreatePredefinedDocs{
+				Docs: getUserDocs(),
+			},
+			testUtils.Request{
+				Request: req,
+				Results: []map[string]any{
+					{"name": "Roy"},
+					{"name": "Addo"},
+					{"name": "Andy"},
+					{"name": "John"},
+					{"name": "Bruno"},
+					{"name": "Chris"},
+					{"name": "Keenan"},
+					{"name": "Shahzad"},
+				},
+			},
+			testUtils.Request{
+				Request:  makeExplainQuery(req),
+				Asserter: testUtils.NewExplainAsserter().WithFieldFetches(8).WithIndexFetches(10),
+			},
+		},
+	}
+
+	testUtils.ExecuteTestCase(t, test)
+}
+
+func TestQueryWithCompositeIndex_WithInFilter_ShouldFetch(t *testing.T) {
+	req := `query {
+		User(filter: {age: {_in: [20, 28, 33]}, name: {_in: ["Addo", "Andy", "Fred"]}}) {
+			name
+		}
+	}`
+	test := testUtils.TestCase{
+		Description: "Test index filtering with _in filter",
+		Actions: []any{
+			testUtils.SchemaUpdate{
+				Schema: schemaWithNameAgeIndex,
+			},
+			testUtils.CreatePredefinedDocs{
+				Docs: getUserDocs(),
+			},
+			testUtils.Request{
+				Request: req,
+				Results: []map[string]any{
+					{"name": "Andy"},
+					{"name": "Fred"},
+				},
+			},
+			testUtils.Request{
+				Request:  makeExplainQuery(req),
+				Asserter: testUtils.NewExplainAsserter().WithFieldFetches(2).WithIndexFetches(3),
+			},
+		},
+	}
+
+	testUtils.ExecuteTestCase(t, test)
+}
+
+func TestQueryWithCompositeIndex_WithNotInFilter_ShouldFetch(t *testing.T) {
+	req := `query {
+		User(filter: {age: {_nin: [20, 23, 28, 42]}, name: {_nin: ["John", "Andy", "Chris"]}}) {
+			name
+		}
+	}`
+	test := testUtils.TestCase{
+		Description: "Test index filtering with _nin filter",
+		Actions: []any{
+			testUtils.SchemaUpdate{
+				Schema: schemaWithNameAgeIndex,
+			},
+			testUtils.CreatePredefinedDocs{
+				Docs: getUserDocs(),
+			},
+			testUtils.Request{
+				Request: req,
```

Comment:
suggestion: I dislike how you had to pull the other super important part of the test out of the action flow, disrupting things for the reader.

I wonder if you could add an action that wraps the two requests perhaps, so the test could read:
```
testUtils.RequestWithExplain{
    Request: `query {
		User(filter: {age: {_nin: [20, 23, 28, 42]}, name: {_nin: ["John", "Andy", "Chris"]}}) {
			name
		}
	}`,
	Results: []map[string]any{
					{"name": "Roy"},
					{"name": "Islam"},
					{"name": "Keenan"},
				},
	Explanation: testUtils.NewExplainAsserter().WithFieldFetches(3).WithIndexFetches(10),
}
``` 
Internally it would split the action into the two Request Actions (that are currently here in the test) before sending them all into the main test executor func.

Happy if you prefer we dont do this in this PR.

Reply:
we've got a task for it https://github.com/sourcenetwork/defradb/issues/1962

Reply:
Ah nice, my bad and thanks :)